### ISSUE TITLE
Single_gpu_test. No attribute data_loader.dataset

### ISSUE THREAD

**Issue by s185231** [open-mmlab/mmdetection#7612]
Hi!
I am very new to mmdetection and have major issues with the toolbox. I am using the SIXray dataset for object detection. Using the demo I have trained the model, saved a checkpoint file and now want to visualize some results (a confusion matrix) . I have imported the single_gpu_test function to create a pickle file. Here is my code:

# Build custom dataset
@DATASETS.register_module()
class SIXray(CustomDataset):
    """XML dataset for detection.

    Args:
        min_size (int | float, optional): The minimum size of bounding
            boxes in the images. If the size of a bounding box is less than
            ``min_size``, it would be add to ignored field.
        img_subdir (str): Subdir where images are stored. Default: JPEGImages.
        ann_subdir (str): Subdir where annotations are. Default: Annotations.
    """
    CLASSES = ('Gun', 'Knife', 'Wrench', 'Pliers', 'Scissors')

    def __init__(self,
                 min_size=None,
                 img_subdir='JPEGImage',
                 ann_subdir='Annotations',
                 **kwargs):

        assert self.CLASSES or kwargs.get(
            'classes', None), 'CLASSES in `SIXray` can not be None.'
        self.img_subdir = img_subdir
        self.ann_subdir = ann_subdir
        super(SIXray, self).__init__(**kwargs)
        self.cat2label = {cat: i for i, cat in enumerate(self.CLASSES)}
        self.min_size = min_size
        self.filter_empty_gt = False

    def load_annotations(self, ann_file):
        """Load annotation from XML style ann_file.

        Args:
            ann_file (str): Path of XML file.

        Returns:
            list[dict]: Annotation info from XML file.
        """

        data_infos = []
        img_ids = mmcv.list_from_file(ann_file)
        """
        for img_id in img_ids:
            filename = osp.join(self.img_subdir, f'{img_id}.jpg')
            print(img_id[0])
            #label_prefix = self.img_prefix.replace('JPEGImage', 'positive-Annotation/Annotation')

            xml_path = osp.join('positive-Annotation/Annotation',
                                f'{img_id}.xml')
            #xml_path = osp.join(self.img_prefix, self.ann_subdir,
            #                    f'{img_id}.xml')
          
            tree = ET.parse(xml_path)
            print(tree)
            root = tree.getroot()
            size = root.find('size')
            if size is not None:
                width = int(size.find('width').text)
                height = int(size.find('height').text)
            else:
                img_path = osp.join(self.img_prefix, filename)
                img = Image.open(img_path)
                width, height = img.size
            data_infos.append(
                dict(id=img_id, filename=filename, width=width, height=height))
        """

        for i in range(1,len(img_ids)):
            if img_ids[i][0] == 'P':
                img_id = img_ids[i][0:6]
            else:
                img_id = img_ids[i][0:8]
            
                #filename = osp.join(self.img_subdir, f'{img_id}.jpg')
            filename = osp.join(f'{img_id}.jpg')

                #label_prefix = self.img_prefix.replace('JPEGImage', 'positive-Annotation/Annotation')
            xml_path = osp.join('/u/data/s194333/SIXray/positive-Annotation/Annotation/',
                                    f'{img_id}.xml')
                #print(filename)
                #print(xml_path)
            #xml_path = osp.join(self.img_prefix, self.ann_subdir,
            #                    f'{img_id}.xml')
          
            tree = ET.parse(xml_path)
            root = tree.getroot()
            size = root.find('size')
            if size is not None:
                width = int(size.find('width').text)
                height = int(size.find('height').text)
            else:
                img_path = osp.join(self.img_prefix, filename)
                img = Image.open(img_path)
                width, height = img.size
            data_infos.append(
                dict(id=img_id, filename=filename, width=width, height=height))
   
        return data_infos

    def get_ann_info(self, idx):
        """Get annotation from XML file by index.

        Args:
            idx (int): Index of data.

        Returns:
            dict: Annotation info of specified index.
        """

        img_id = self.data_infos[idx]['id']
        xml_path = osp.join('/u/data/s194333/SIXray/positive-Annotation/Annotation/',
                                f'{img_id}.xml')

        #xml_path = osp.join(self.img_prefix, self.ann_subdir, f'{img_id}.xml')
        tree = ET.parse(xml_path)
        root = tree.getroot()
        bboxes = []
        labels = []
        bboxes_ignore = []
        labels_ignore = []
        for obj in root.findall('object'):
            name = obj.find('name').text
            if name not in self.CLASSES:
                continue
            label = self.cat2label[name]
            #difficult = obj.find('difficult')
            #difficult = 0 if difficult is None else int(difficult.text)
            bnd_box = obj.find('bndbox')
            # TODO: check whether it is necessary to use int
            # Coordinates may be float type
            bbox = [
                int(float(bnd_box.find('xmin').text)),
                int(float(bnd_box.find('ymin').text)),
                int(float(bnd_box.find('xmax').text)),
                int(float(bnd_box.find('ymax').text))
            ]
            ignore = False
            if self.min_size:
                assert not self.test_mode
                w = bbox[2] - bbox[0]
                h = bbox[3] - bbox[1]
                if w < self.min_size or h < self.min_size:
                    ignore = True
            #if difficult or ignore:
            #    bboxes_ignore.append(bbox)
            #    labels_ignore.append(label)
            else:
                bboxes.append(bbox)
                labels.append(label)
        if not bboxes:
            bboxes = np.zeros((0, 4))
            labels = np.zeros((0, ))
        else:
            bboxes = np.array(bboxes, ndmin=2) - 1
            labels = np.array(labels)
        if not bboxes_ignore:
            bboxes_ignore = np.zeros((0, 4))
            labels_ignore = np.zeros((0, ))
        else:
            bboxes_ignore = np.array(bboxes_ignore, ndmin=2) - 1
            labels_ignore = np.array(labels_ignore)
        ann = dict(
            bboxes=bboxes.astype(np.float32),
            labels=labels.astype(np.int64),
            bboxes_ignore=bboxes_ignore.astype(np.float32),
            labels_ignore=labels_ignore.astype(np.int64))
        return ann

    def get_cat_ids(self, idx):
        """Get category ids in XML file by index.

        Args:
            idx (int): Index of data.

        Returns:
            list[int]: All categories in the image of specified index.
        """

        cat_ids = []
        img_id = self.data_infos[idx]['id']
        xml_path = osp.join('/u/data/s194333/SIXray/positive-Annotation/Annotation/',
                            f'{img_id}.xml')
        
        tree = ET.parse(xml_path)
        root = tree.getroot()
        for obj in root.findall('object'):
            name = obj.find('name').text
            if name not in self.CLASSES:
                continue
            label = self.cat2label[name]
            cat_ids.append(label)

        return cat_ids

config_file = '/u/data/s185231/mmdetection-master/configs/faster_rcnn/faster_rcnn_r50_caffe_dc5_mstrain_1x_coco.py'
checkpoint = 'my_checkpoint/epoch_12.pth'
device='cuda:5'

# Load the config
config = mmcv.Config.fromfile(config_file)
config.dataset_type = 'SIXray'
config.data_root = '/u/data/s194333/SIXray'
config.data.test.type = 'SIXray'
config.data.test.data_root = '/u/data/s194333/SIXray/'
config.data.test.ann_file = 'TestTrainSplits/Test_05.csv'
config.data.test.img_prefix = 'JPEGImage'

config.data.train.type = 'SIXray'
config.data.train.data_root = '/u/data/s194333/SIXray/'
config.data.train.ann_file = 'TestTrainSplits/Train_05.csv'
config.data.train.img_prefix = 'JPEGImage'

config.data.val.type = 'SIXray'
config.data.val.data_root = '/u/data/s194333/SIXray/'
config.data.val.ann_file = 'TestTrainSplits/Val_05.csv'
config.data.val.img_prefix = 'JPEGImage'

config.model.roi_head.bbox_head.num_classes = 5
config.model.pretrained = None

# Initialize the detector
model = build_detector(config.model, test_cfg=config.get('test_cfg'))
model.cfg = config
# Load checkpoint
checkpoint = load_checkpoint(model, checkpoint, map_location=device)

# Test the model
model.CLASSES = checkpoint['meta']['CLASSES']
model.to(device)
model.eval()
tests = build_dataset(model.cfg.data.test)
config.work_dir = './results'
mmcv.mkdir_or_exist(osp.abspath(config.work_dir))
show_score_thr=0.1
single_gpu_test(model,tests,config.work_dir,show_score_thr)

# Error 
dataset = data_loader.dataset - Attribute error
What am I doing wrong?

**Comment by chhluo** [comment]
tests should be a dataloader object rather than dataset object, see https://github.com/open-mmlab/mmdetection/blob/3e2693151add9b5d6db99b944da020cba837266b/tools/test.py#L208
https://github.com/open-mmlab/mmdetection/blob/3e2693151add9b5d6db99b944da020cba837266b/tools/test.py#L233
